### PR TITLE
`ultralytics 8.3.237` SAM3 integration

### PR DESCRIPTION
**Pull request by Laughing-q**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
Adds full SAM 3 integration (image + video, interactive + semantic) with updated docs, new builders, and predictor APIs—plus several low-level SAM/SAM2 refactors and quality fixes 🚀🧩

### 📊 Key Changes
- **Docs overhaul for SAM 3 usage** 📝  
  - Clarifies that **SAM 3 weights (`sam3.pt`) are not auto-downloaded** and must be manually fetched from the official repo.  
  - Adds a requirement note for **BPE vocab** (`bpe_simple_vocab_16e6.txt.gz`) for text prompts.  
  - Replaces “API preview” examples with **concrete predictor-based examples** (semantic text prompts, exemplar boxes, feature reuse, video tracking).
- **New SAM 3 model builder module** 🏗️  
  - Adds `ultralytics/models/sam/build_sam3.py` with builders for:
    - **Semantic image model** (`build_sam3_image_model`) requiring `bpe_path`
    - **Interactive/video tracking model** (`build_interactive_sam3`)
- **New SAM 3 predictors and exports** 🎯  
  - Adds and exports: `SAM3Predictor`, `SAM3SemanticPredictor`, `SAM3VideoPredictor`, `SAM3VideoSemanticPredictor`.  
  - Updates `SAM` model selection to route `sam3*.pt` to the right predictor/model automatically.
- **Core predictor + stride handling improvements** ⚙️  
  - `BasePredictor.setup_source()` now accepts an optional `stride` argument to validate `imgsz` correctly for models like SAM 3 (stride 14).  
  - SAM predictors now use `self.imgsz` consistently (instead of `self.args.imgsz`) and properly set model image size.
- **Checkpoint loading refactor for SAM/SAM2** 📦  
  - Introduces a shared `_load_checkpoint()` in `build.py` handling both flat and nested `"model"` checkpoints.
- **SAM3 + attention/encoder utilities updated** 🔧  
  - Uses PyTorch `scaled_dot_product_attention` in a key attention path (performance/clarity win).  
  - Adds mask interpolation support in `MaskDownSampler`, flexible custom attention modules in `MemoryAttentionLayer`, and new position/relative-pos helpers in utils.
- **Expanded API reference documentation** 📚  
  - Adds many new reference pages for SAM 3 internals (encoder/decoder/tokenizer/etc.) and updates MkDocs navigation.

### 🎯 Purpose & Impact
- **Makes SAM 3 реально usable in Ultralytics** ✅  
  Users can run SAM 3 for:
  - **Concept segmentation with text prompts** (semantic predictor + BPE vocab) 🧠🔤  
  - **Exemplar/box-driven concept segmentation** 🧰  
  - **Video tracking (interactive + semantic)** 🎥🔍
- **Reduces “it doesn’t work” setup confusion** ⚠️➡️✅  
  Explicitly documents that SAM 3 requires **manual weight download** and (for text) a **separate BPE file**, preventing silent failures.
- **Better correctness for SAM 3 image sizing** 📐  
  Stride-aware `imgsz` handling avoids mis-sized preprocessing and improves reliability across SAM/SAM2/SAM3.
- **More flexible & efficient workflows** 🏎️  
  The docs and APIs now highlight **feature reuse** (extract once, query many), which can significantly speed up repeated queries on the same image.
- **Foundation for richer SAM 3 ecosystem in Ultralytics** 🧱  
  The new builders, predictors, and reference docs pave the way for future polish (CLI parity, smoother asset handling, broader examples, etc.).

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review helper; a Ultralytics engineer will be with you shortly. In the meantime, please review the checklist below to help ensure a smooth integration:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the **Update branch** button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features. This PR already adds substantial SAM 3 docs; please double-check for completeness and consistency.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing `I have read the CLA Document and I sign the CLA` in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have any questions about the SAM 3 integration, API design, or docs, feel free to leave a comment. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/sam/predict.py [hunk 1]
```diff
@@ -1986,3 +2152,1797 @@ def track_step(
...
+            pred_masks, pred_boxes = None, torch.zeros((0, 7), device=pred_masks.device)
+        else:
+            pred_masks = torch.cat([obj_id_to_mask[obj_id] for obj_id in curr_obj_ids], dim=0)
+            pred_masks = F.interpolate(pred_masks.float()[None], orig_imgs[0].shape[:2], mode="bilinear")[0] > 0.5
+            pred_ids = torch.tensor(curr_obj_ids, dtype=torch.int32, device=pred_masks.device)
+            pred_scores = torch.tensor(
+                [preds["obj_id_to_score"][obj_id] for obj_id in curr_obj_ids], device=pred_masks.device
+            )
+            pred_cls = torch.tensor(
+                [preds["obj_id_to_cls"][obj_id] for obj_id in curr_obj_ids], device=pred_masks.device
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py:2636]: ⚠️ **HIGH**: When no objects are returned, this branch tries to build `pred_boxes` with `pred_masks.device` even though `pred_masks` has not been defined yet, which immediately raises an `UnboundLocalError`. Even after that, the code below continues to access `pred_masks` (e.g., `pred_masks.shape[0]`) despite it being `None`, so the whole `postprocess` path crashes whenever the predictor outputs zero tracks. Please create the empty tensor on a known device (e.g., `self.device`) and exit the method early or guard the remaining logic when no masks are available.

**Suggested change:**
```suggestion
        return [Results(orig_imgs[0], path=self.batch[0][0], names={}, masks=None, boxes=torch.zeros((0, 7), device=self.device))]
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Two blocking issues emerged in the SAM3 video semantic path: (1) `SAM3VideoSemanticPredictor.postprocess` tries to reference `pred_masks` before it exists and continues to treat it as a tensor even when no objects are predicted, so zero-output frames always crash; guard that branch and return safely. (2) The IoU-based reconditioning flow calls `batched_mask_to_box` without importing or defining it, which raises a `NameError` as soon as the feature is enabled; bring the utility into scope or remove the call.

💬 Posted 2 inline comments

⚠️ **Large PR**: Review focused on critical issues. Some details may not be covered.

File: ultralytics/models/sam/predict.py [hunk 2]
```diff
@@ -1986,3 +2152,1797 @@ def track_step(
...
+
+                tracker_mask = tracker_low_res_masks_global[trk_idx]
+                mask_binary = tracker_mask > 0
+                mask_area = mask_binary.sum().item()
+
+                if mask_area == 0:
+                    continue  # Skip tracklets with zero mask area
+
+                # Get bounding box from SAM2 mask and convert to normalized coordinates
+                tracker_box_pixels = batched_mask_to_box(mask_binary.unsqueeze(0)).squeeze(0)
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py:3136]: ⚠️ **HIGH**: `batched_mask_to_box` is called here but never imported or defined in this module, so enabling `reconstruction_bbox_iou_thresh` immediately raises a `NameError`. Import the helper (e.g., `from ultralytics.utils.ops import batched_mask_to_box`) or replace the call with an available utility before referencing it.

**Suggested change:**
```suggestion
                from ultralytics.utils.ops import batched_mask_to_box
```

**Comment by glenn-jocher** [comment]
Maybe we need a min Python version for SAM3?

**Comment by Jordan-Pierce** [comment]
> Maybe we need a min Python version for SAM3?

Regarding images, I was able to run the following without needing to update from `3.10`
```python
from ultralytics import SAM

model = SAM("sam3.pt")
# Segment with point(s)
results = model("zidane.jpg", points=[900, 370], labels=[1])[0].show()
# Segment everything (if no prompts are provided?)
model("zidane.jpg")[0].show()
```

But the following required addition dependencies (`triton`, etc.) and upgrading to `3.12`
```python
from ultralytics.models.sam import SAM3Predictor, SAM3SemanticPredictor

# Create SAMPredictor
overrides = dict(conf=0.25, task="segment", mode="predict", imgsz=1024, model="sam3.pt")
predictor = SAM3Predictor(overrides=overrides)

# Set image
predictor.set_image("zidane.jpg")  # set with image file

# Run inference with single point prompt
results = predictor(points=[900, 370], labels=[1])[0].show()
```

**Comment by Laughing-q** [comment]
@glenn-jocher not sure if this is related to the python version here, I just tried to do a local python3.8 test and it passed
pic-251209-1117-05

**Comment by glenn-jocher** [comment]
Trying slow tests here



## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2025-12-10 01:19:09 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/20083986666): `ci.yml` run 15296
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/20083986796): `docker.yml` run 1487

**Comment by Laughing-q** [comment]
@glenn-jocher It turned out it's because there are some `torch.inference_mode` introduced in this PR and it's not compatible with torch==1.8.0 in our python=3.8 test env. And because of the lazy import mechanism in our `ultralytics/__init__.py`, the error did explicitly show up, and shown import error instead.
The local test I did which passed, were not using torch==1.8.0 so did not spot that.
Now tests are all good! Except the tflite export failure in slow test, which I think it's not relevant this PR. We can fix that in another PR probably.

**Comment by UltralyticsAssistant** [comment]
Thrilled to see this PR merged — huge congratulations @Laughing-q, and heartfelt thanks to @Jordan-Pierce and @glenn-jocher for the support and collaboration. 🚀

As Leonardo da Vinci said, **“Simplicity is the ultimate sophistication.”** This merge brings that spirit to life by making SAM 3 *practically usable* in Ultralytics: clear docs that prevent setup gotchas (manual `sam3.pt` + BPE vocab), solid new builders (`build_sam3.py`), and a clean predictor lineup (image + video, interactive + semantic) that routes `sam3*.pt` automatically. The stride-aware `imgsz` handling and checkpoint-loading refactors are especially impactful—quiet improvements that save a lot of user time and confusion.

Thank you all for pushing the ecosystem forward—this lays a strong foundation for a richer SAM 3 experience across YOLO11-era workflows and beyond.

**Comment by glenn-jocher** [comment]
Merged, nice work!!
